### PR TITLE
cmake: Fix generated header file dependency

### PR DESCRIPTION
**Pull request by joerchan**
Fix race condition in build when creating libraries in application and
depending on zephyr_interface.

lib/CMakeLists.txt:
```
    add_library(lib STATIC src/lib.c)
    zephyr_append_cmake_library(lib)
    target_link_libraries(lib PUBLIC zephyr_interface)
```
Gives following error message:
```
  from path/src/lib.c:6:
  zephyr/include/syscall.h:11:10: fatal error: syscall_list.h: \
  	No such file or directory
     11 | #include <syscall_list.h>
        |          ^~~~~~~~~~~~~~~~
  compilation terminated.
  [13/165] Generating include/generated/syscall_dispatch.c, \
  	include/generated/syscall_list.h
```

### CODE REVIEW

**Comment by tejlmand** [comment]
> ```
>     add_library(lib STATIC src/lib.c)
>     zephyr_append_cmake_library(lib)
>     target_link_libraries(lib PUBLIC zephyr_interface)
> ```

when depending on generated headers, then targets should add a dependency to: `zephyr_generated_headers`.

That is:
```
add_library(lib STATIC src/lib.c)
zephyr_append_cmake_library(lib)
target_link_libraries(lib PUBLIC zephyr_interface)
add_dependencies(lib zephyr_generated_headers)  
```
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L652-L655

unfortunately it seems @SebastianBoe never got the time to update the docs.: https://github.com/zephyrproject-rtos/zephyr/pull/22143#issuecomment-577763234

Will need a little time to fully ensure that changing dependencies has no side-effects.
Changed to enhancment, as this is not a bug.

**Comment by tejlmand** [comment]
`DNM` until I have had time to completely ensure there are no side-effects to changing the dependencies.

**Comment by joerchan** [comment]
@tejlmand Note that in this case it wasn't me that depended on the generated the header files, it is the zephyr headers itself.

I got the same issue when using zephyrs own wrappers to create header (because they exactly the same).

```
zephyr_library_named(lib)
zephyr_library_sources(lib_src.c)
```

I have seen the question about how to properly create libraries in the app and link them in the zephyr build system pop up many places.
The samples only have externally built static library, I think adding application libraries sample would be helpful.

**Comment by tejlmand** [comment]
@joerchan thanks for a little more details.

**Note** using Zephyr's macros `zephyr_library` AFTER `find_package(Zephyr)` also has limitations.
One limitation is that libraries created after `find_package(Zephyr)` will not be included inside the `--whole-archive` linker flags, and this may cause unintended side effects for the user.
So if wanting to use `zephyr_library` macros, then it is recommended to get the library included into the build system as a Zephyr module.

This limitation on Zephyr macros is something I also would like to improve in the future.

**Comment by tejlmand** [comment]
> I have seen the question about how to properly create libraries in the app and link them in the zephyr build system pop up many places.
> The samples only have externally built static library, I think adding application libraries sample would be helpful.

I agree. The guides has too little information on how to properly create libs outside Zephyr itself.
Especially when taken into consideration the limitations when creating a lib as part of the app.

**Comment by joerchan** [comment]
> DNM until I have had time to completely ensure there are no side-effects to changing the dependencies.

Did you have any chance to look at this?

If you have a task for static library improvements, then link order of libgcc could be added as well.
Had another issue where -lgcc had to be added after the extra library to resolve link errors.

**Comment by tejlmand** [comment]
> > DNM until I have had time to completely ensure there are no side-effects to changing the dependencies.
> 
> Did you have any chance to look at this?

Unfortunately not.

> 
> If you have a task for static library improvements, then link order of libgcc could be added as well.
> Had another issue where -lgcc had to be added after the extra library to resolve link errors.

I believe that is related to:
https://github.com/zephyrproject-rtos/zephyr/issues/28650

**Comment by joerchan** [comment]
> I believe that is related to:

Seems like the same fix atleast, we had failure to link with libgcc in the last library in the link command. Because -lgcc was before it.
Adding it both before and after worked.

**Comment by Mattemagikern** [comment]
Hello, 
I've just tested this patch since I ran into the same bulid error as described and it works very well.
With the patch I got consistent bulids with external libraries. 

I wolud like this patch to be merged :+1: 

Best regards,

**Review by tyrantlizard77 — APPROVED** [review]
My CMake-fu is weak but I remember doing something similar for a build issue with some userspace stuff.

@tejlmand can you take another look?

File: CMakeLists.txt [hunk 1]
```diff
@@ -670,12 +671,12 @@ target_include_directories(${OFFSETS_LIB} PRIVATE
...
   ${ARCH_DIR}/${ARCH}/include
   )
-target_link_libraries(${OFFSETS_LIB} zephyr_interface)
-add_dependencies(     ${OFFSETS_LIB}
+add_dependencies(zephyr_interface
   ${SYSCALL_LIST_H_TARGET}
   ${DRIVER_VALIDATION_H_TARGET}
   ${KOBJ_TYPES_H_TARGET}
   )
+target_link_libraries(${OFFSETS_LIB} zephyr_interface)
```
Comment (tejlmand) [inline CMakeLists.txt:679]: No need to move this line.
Can stay together with the other CMake calls handling `${OFFSETS_LIB}`

**Review by tejlmand — APPROVED** [review]
Sorry for the long response time.
I have taken a tour in the CMake dependencies, and this seems safe to merge.

Thanks for the fix.

File: CMakeLists.txt (hunk 1 above)
Comment (joerchan) [reply CMakeLists.txt]: Reverted so that line is not moved.

**Comment by joerchan** [comment]
@tejlmand Rebased and fixed your comment.

Can you remove the DNM label?

**Comment by tejlmand** [comment]
> Can you remove the DNM label?

Done.

File: CMakeLists.txt [hunk 2]
```diff
@@ -566,9 +566,12 @@ add_custom_command(
   DEPENDS ${syscalls_subdirs_trigger} ${PARSE_SYSCALLS_HEADER_DEPENDS}
   )
 
-add_custom_target(${SYSCALL_LIST_H_TARGET} DEPENDS             ${syscall_list_h})
+add_custom_target(${SYSCALL_LIST_H_TARGET} DEPENDS ${syscall_list_h})
 add_custom_target(${PARSE_SYSCALLS_TARGET}
-	          DEPENDS ${syscalls_json} ${struct_tags_json})
+  DEPENDS
+  ${syscalls_json}
+  ${subsys_jsonstruct_tags_json}
```
Comment (nashif) [inline CMakeLists.txt:573]: where is this guy coming from?

Comment (joerchan) [reply CMakeLists.txt]: @nashif Honestly, no clue, but wrong.

Comment (nashif) [reply CMakeLists.txt]: :)

**Pull request merged**